Project.toml

```diff
@@ -1,7 +1,7 @@
 name = "Distributions"
 uuid = "31c24e10-a181-5473-b8eb-7969acd0382f"
 authors = ["JuliaStats"]
-version = "0.25.90"
+version = "0.25.91"
```

Comment:
I did a JuliaHub package search to see whether any packages might be relying on `Truncated{D,S}` being concrete and/or using the bounds directly and assuming they're real-valued. Looks like we're in the clear for the former but there are a few registered packages that will break due to the latter: [ModelConstructors](https://github.com/FRBNY-DSGE/ModelConstructors.jl/blob/main/src/distributions_ext.jl#L440C36-L458), [DiffEqBayes](https://github.com/SciML/DiffEqBayes.jl/blob/master/src/stan_string.jl#L210-L215), and [DiffEqBayesStan](https://github.com/StanJulia/DiffEqBayesStan.jl/blob/master/src/stan_string.jl#L210-L215).

Reply:
I would argue that this is an internal change and things may break for downstream packages that access fields of (to be internal) structs that are not part of the API.

Reply:
For instance, I don't think code such as in ModelConstructors is a reason to move this change to a breaking release.

Reply:
> access fields of (to be internal) structs

? I'm not sure what you mean by "to be internal" here.

I am generally of the opinion that if you touch the fields/properties of an object then you should expect to lose a finger at any moment. That said, it doesn't seem to be documented that these fields aren't public, though the fields themselves also aren't documented so perhaps it can be reasonably assumed that they aren't public.

> I don't think code such as in ModelConstructors is a reason to move this change to a breaking release.

That's fair, but it could be worthwhile to submit PRs to the packages that will be affected by this change to ensure they use `extrema` (or similar) instead to avoid any issues.

Reply:
> ? I'm not sure what you mean by "to be internal" here.

I meant that generally I think these fields do not belong to any official API, so there are no guarantees at all and code that accesses them may break at any time, but that in this case even the whole struct is intended to be an internal implementation detail in the next breaking release. For backward compatibility, `Truncated` is still exported but all outer constructors are already deprecated since users are supposed to call `truncated` instead.

Reply:
> I meant that generally I think these fields do not belong to any official API, so there are no guarantees at all and code that accesses them may break at any time

👍

> but that in this case even the whole struct is intended to be an internal implementation detail in the next breaking release

🤨 I was definitely not aware of this. Where was that discussed/decided?

Reply:
`Truncated` has been deprecated for a long time: https://github.com/JuliaStats/Distributions.jl/pull/1026 This change was also reflected in the docs which state that you should use `truncated`.

Reply:
But it has not been discussed that the deprecations are removed in exactly "the next breaking release", more precise would have been "a breaking release".

Reply:
Using the type constructor as the means of doing truncation is deprecated, not the actual binding itself. As I understood that change, the motivation was to provide an interface for truncating distributions in a way that could produce something other than a `Truncated`-wrapped distribution when there is a more suitable choice, e.g. `truncated(LogNormal(); lower=0)` could produce a `LogNormal` rather than a `Truncated{LogNormal}`. AFAICT there's been no mention of removing `Truncated` from the API altogether; it's even referenced by the `truncated` docstring as the fallback return type. I don't see why it would be desirable to make it internal, plus it would be quite annoying for anyone defining methods for it.

Reply:
I think probably a middle ground would be good: Not exporting `Truncated` anymore (since as long as it is exported, users will construct truncated distributions with available constructors of `Truncated` even though it is deprecated and not recommended) but mentioning it in the docs as a fallback that can be dispatched on without any guarantees on the fields or type parameters.

Reply:
Regarding your initial comment, I submitted PRs to the three affected packages: https://github.com/SciML/DiffEqBayes.jl/pull/304, https://github.com/StanJulia/DiffEqBayesStan.jl/pull/18, https://github.com/FRBNY-DSGE/ModelConstructors.jl/pull/18

Reply:
Once the deprecated constructor methods are removed, the only remaining method for `Truncated` is its inner constructor. If a user really wants to go through the trouble of computing all of the requisite quantities that get stored in `Truncated` in order to pass them to the inner constructor, they either have a very good reason to do so or all hope is lost and they're simply an agent of chaos who will do as they please regardless of whether it's exported.

I'd recommend we continue to export `Truncated` but add a note to its docstring that it isn't intended to be constructed directly and to use `truncated` for that purpose (currently `truncated` links to `Truncated` but not vice versa), and that the fields are an implementation detail. The one guarantee it would be nice to make is that the first type parameter is a `Distribution` but there may be more so don't assume `Truncated{D}` is concrete.

Reply:
But why export it if you should not use it? To me it seems more reasonable to not export it and only clearly state and document the guarantees on the fallback type (and hence make changes to them require a breaking release).

Reply:
> But why export it if you should not use it?

IMO it's not that you shouldn't _use_ it, just that you shouldn't assume that its constructor is The Way™ to obtain a truncated distribution.